### PR TITLE
finish week2 progress

### PR DESCRIPTION
**Pull request by coookie89**
### Todo
- [x] ch4
- [x] ch5

### CODE REVIEW

File: coookie89/Week2/ch4/4.2/exercise1.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            torch.randn(num_hidden_test, num_outputs, requires_grad=True) * 0.01
+        )
+        b2 = nn.Parameter(torch.zeros(num_outputs, requires_grad=True))
+
+        params = [W1, b1, W2, b2]
+
+        loss = nn.CrossEntropyLoss()
+
+        num_epochs = 10
+        lr = 0.1  # learning rate
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.2/exercise1.py:46]: 可以放到上面變常數，其實 batch_size 還有前面的 input output 也可以是常數就是。

File: coookie89/Week2/ch4/4.2/exercise1.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        )
+        b1 = nn.Parameter(torch.zeros(num_hidden_test, requires_grad=True))
+        W2 = nn.Parameter(
+            torch.randn(num_hidden_test, num_outputs, requires_grad=True) * 0.01
+        )
+        b2 = nn.Parameter(torch.zeros(num_outputs, requires_grad=True))
+
+        params = [W1, b1, W2, b2]
+
+        loss = nn.CrossEntropyLoss()
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.2/exercise1.py:43]: 這邊改叫 loss_function 會比較好，loss 會讓人誤會他是被計算出來的 loss

Comment (jamfly) [reply coookie89/Week2/ch4/4.2/exercise1.py:43]: 甚至不用給個名字，直接塞進 function 裡面都是可以接受的

File: coookie89/Week2/ch4/4.3/exercise2.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    nets.append(
+        nn.Sequential(
+            nn.Flatten(), nn.Linear(784, 256), nn.Sigmoid(), nn.Linear(256, 10)
+        )
+    )
+    nets.append(
+        nn.Sequential(
+            nn.Flatten(), nn.Linear(784, 256), nn.Tanh(), nn.Linear(256, 10)
+        )
+    )
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.3/exercise2.py:30]: 可以寫成 function

File: coookie89/Week2/ch4/4.4/example.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+# 4.4.4.2. Training and Testing the Model
+
+
+def evaluate_loss(net, data_iter, loss):
+    """Evaluate the loss of a model on the given dataset."""
+    metric = d2l.Accumulator(2)  # Sum of losses, no. of examples
+    for X, y in data_iter:
+        out = net(X)
+        y = y.reshape(out.shape)
+        losses = loss(out, y)
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.4/example.py:19]: loss 改成 loss_function 比較可讀

File: coookie89/Week2/ch4/4.4/example.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    # Switch off the bias since we already catered for it in the polynomial features
+    net = nn.Sequential(nn.Linear(input_shape, 1, bias=False))
+    batch_size = min(10, train_labels.shape[0])
+    train_iter = d2l.load_array(
+        (train_features, train_labels.reshape(-1, 1)), batch_size
+    )
+    test_iter = d2l.load_array(
+        (test_features, test_labels.reshape(-1, 1)), batch_size, is_train=False
+    )
+    trainer = torch.optim.SGD(net.parameters(), lr=0.01)
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.4/example.py:38]: 這邊的 lr 也可以變成參數

File: coookie89/Week2/ch4/4.5/example.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    print("L2 norm of w:", torch.norm(w).item())
+
+
+def train_concise(wd):
+    """Concise Implementation"""
+    net = nn.Sequential(nn.Linear(num_inputs, 1))
+    for param in net.parameters():
+        param.data.normal_()
+    loss = nn.MSELoss()
+    num_epochs, lr = 100, 0.003
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.5/example.py:57]: 改成參數吃進來會比較乾淨

**Review by jamfly — CHANGES_REQUESTED** [review]
一些重複出現的小地方需要修正

File: coookie89/Week2/ch4/4.2/exercise1.py (hunk 1 above)
Comment (coookie89) [reply coookie89/Week2/ch4/4.2/exercise1.py:46]: 是指放在main的最上面嗎
還是把它放在整份程式的最上面(當全域變數)

Comment (coookie89) [reply coookie89/Week2/ch4/4.2/exercise1.py:46]: 是改成THIS_IS_THE_num_epochs = 10
然後放程式最上面嗎！

Comment (jamfly) [reply coookie89/Week2/ch4/4.2/exercise1.py:46]: `NUM_EPOCHS = 10`

File: coookie89/Week2/ch4/4.2/exercise1.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+import torch
+from torch import nn
+from d2l import torch as d2l
+
+BATCH_SIZE = 256
+
+NUM_INPUTS = 784  # 28*28 pixels
+NUM_OUTPUTS = 10  # 分成10個類別
+
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.2/exercise1.py:12]: 不要有空格，所有常數都排在一起

File: coookie89/Week2/ch4/4.2/exercise1.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+def net(X):
+    X = X.reshape((-1, NUM_INPUTS))
+    H = relu(X @ W1 + b1)  # Here '@' stands for matrix multiplication
+    return H @ W2 + b2
+
+
+if __name__ == "__main__":
+
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.2/exercise1.py:30]: 不要有空格

File: coookie89/Week2/ch4/4.2/exercise2.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+
+if __name__ == "__main__":
+    batch_size = 256
+    train_iter, test_iter = d2l.load_data_fashion_mnist(batch_size)
+
+    num_inputs = 784  # 28*28 pixels
+    num_outputs = 10  # 分成10個類別
+    num_hiddens_1 = 256
+    num_hiddens_2 = 128
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.2/exercise2.py:28]: 跟上面的一樣可以改成常數

File: coookie89/Week2/ch4/4.2/exercise2.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    W1 = nn.Parameter(
+        torch.randn(num_inputs, num_hiddens_1, requires_grad=True) * 0.01
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.2/exercise2.py:31]: 不要用魔術數字 0.1，把 0.1 變成一個變數

File: coookie89/Week2/ch4/4.3/exercise1.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def init_weights(m):
+    if type(m) == nn.Linear:
+        nn.init.normal_(m.weight, std=0.01)
+
+
+if __name__ == "__main__":
+
+    # 兩層hidden layers
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.3/exercise1.py:15]: 不要有空格

File: coookie89/Week2/ch4/4.3/exercise1.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        nn.Linear(256, 64),
+        nn.ReLU(),
+        nn.Linear(64, 10),
+    )
+    net.apply(init_weights)
+
+    batch_size = 256
+    lr = 0.5
+    num_epochs = 10
+
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.3/exercise1.py:29]: 常數

File: coookie89/Week2/ch4/4.3/exercise2.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    return nn.Sequential(
+        nn.Flatten(),
+        nn.Linear(784, 256),
+        activation_function,
+        nn.Linear(256, 10),
+    )
+
+
+if __name__ == "__main__":
+
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.3/exercise2.py:23]: 不要有空格

File: coookie89/Week2/ch4/4.3/exercise2.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    nets = []
+    nets.append(apply_activation_functions(nn.ReLU()))
+    nets.append(apply_activation_functions(nn.Sigmoid()))
+    nets.append(apply_activation_functions(nn.Tanh()))
+
+    batch_size = 256
+    lr = 0.3
+    num_epochs = 10
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.3/exercise2.py:31]: 常數

File: coookie89/Week2/ch4/4.3/exercise2.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+    train_iter, test_iter = d2l.load_data_fashion_mnist(batch_size)
+
+    for net_id in range(0, 3):
+
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.3/exercise2.py:36]: 不要有空格

File: coookie89/Week2/ch4/4.3/exercise2.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        if net_id == 0:
+            print("using ReLU function\n")
+        elif net_id == 1:
+            print("using Sigmoid function\n")
+        elif net_id == 2:
+            print("using Tanh function\n")
+
```
Comment (jamfly) [inline coookie89/Week2/ch4/4.3/exercise2.py:43]: 像這種情況 建議把 else 也寫完整，有時候會出現沒想到的 case，這可以幫助 debug
```
else:
    print(f"{net_id=}")
```

**Review by jamfly — CHANGES_REQUESTED** [review]

**Review by jamfly — APPROVED** [review]

**Comment by jamfly** [comment]
README 忘了寫日期，但已經 merge 了，下一條 PR 再補吧
